### DISCUSSION TITLE
RTSP runOnDemand keeps Gstreamer source running until 60s session timeout

### DISCUSSION THREAD

**Discussion by TrueDenRyu** [bluenviron/mediamtx#3109] [General]
### Question

Greetings,

I'm new to MediaMTX but impressed with what I've seen with its functionality and community support. I'm currently observing an issue on my custom Onvif image sensor setup. I'm currently using the built release file v1.6.0_linux_arm64v8 (I was also having the same issue on v1.5.1 that I built from source).  I'm using AgentDVR crossplatform software to capture the steam from an embedded system running Yocto Kirkstone Linux that has the image sensor attached. 

There are a few hardware related plugins in my gstreamer command but the runOnDemand is as follows:
`runOnDemand: gst-launch-1.0 v4l2src io-mode=dmabuf-import device=/dev/video3 ! video/x-bayer, width=2600, height=1952, format=grbg10 !  tiovxisp sink_0::device=/dev/v4l-subdev2 sensor-name=SENSOR_ONSEMI_AR0521 dcc-isp-file=/opt/imaging/ar0521/linear/dcc_viss_GRBG_10b_2600x1952.bin sink_0::dcc-2a-file=/opt/imaging/ar0521/linear/dcc_2a_GRBG_10b_2600x1952.bin format-msb=9 sink_0::pool-size=8 src_0::pool-size=8 ! queue ! v4l2h264enc ! rtspclientsink location=rtsp://localhost:554/test
`
Here is the output of my server log:
```
2024/02/25 00:13:12 INF MediaMTX v1.6.0
2024/02/25 00:13:12 INF configuration loaded from /opt/onvif/mediamtx.yml
2024/02/25 00:13:12 DEB path manager created
2024/02/25 00:13:12 INF [RTSP] listener opened on :554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
2024/02/25 00:13:12 INF [RTMP] listener opened on :1935
2024/02/25 00:13:12 INF [HLS] listener opened on :8888
2024/02/25 00:13:12 INF [WebRTC] listener opened on :8889 (HTTP), :8189 (ICE/UDP)
2024/02/25 00:13:12 INF [SRT] listener opened on :8890 (UDP)
2024/02/25 00:13:16 INF [RTSP] [conn 192.168.10.133:58145] opened
2024/02/25 00:13:16 DEB [RTSP] [conn 192.168.10.133:58145] [c->s] OPTIONS rtsp://192.168.10.122:554/test RTSP/1.0
CSeq: 1
User-Agent: Mozilla/5.0


2024/02/25 00:13:16 DEB [RTSP] [conn 192.168.10.133:58145] [s->c] RTSP/1.0 200 OK
CSeq: 1
Public: DESCRIBE, ANNOUNCE, SETUP, PLAY, RECORD, PAUSE, GET_PARAMETER, TEARDOWN
Server: gortsplib


2024/02/25 00:13:16 DEB [RTSP] [conn 192.168.10.133:58145] [c->s] DESCRIBE rtsp://192.168.10.122:554/test RTSP/1.0
Accept: application/sdp
CSeq: 2
User-Agent: Mozilla/5.0


2024/02/25 00:13:16 DEB [path test] created
2024/02/25 00:13:16 INF [path test] runOnDemand command started
APP: Init ... !!!
MEM: Init ... !!!
MEM: Initialized DMA HEAP (fd=6) !!!
MEM: Init ... Done !!!
IPC: Init ... !!!
IPC: Init ... Done !!!
REMOTE_SERVICE: Init ... !!!
REMOTE_SERVICE: Init ... Done !!!
  5680.123138 s: GTC Frequency = 200 MHz
APP: Init ... Done !!!
  5680.130961 s:  VX_ZONE_INIT:Enabled
  5680.131191 s:  VX_ZONE_ERROR:Enabled
  5680.131378 s:  VX_ZONE_WARNING:Enabled
  5680.132723 s:  VX_ZONE_INIT:[tivxInitLocal:130] Initialization Done !!!
  5680.134418 s:  VX_ZONE_INIT:[tivxHostInitLocal:96] Initialization Done for HOST !!!
Setting pipeline to PAUSED ...
Pipeline is live and does not need PREROLL ...
Progress: (open) Opening Stream
Progress: (connect) Connecting to rtsp://localhost:554/test
Pipeline is PREROLLED ...
Prerolled, waiting for progress to finish...
2024/02/25 00:13:17 INF [RTSP] [conn [::1]:34110] opened
Progress: (open) Retrieving server options
2024/02/25 00:13:17 DEB [RTSP] [conn [::1]:34110] [c->s] OPTIONS rtsp://localhost:554/test RTSP/1.0
CSeq: 1
Date: Sun, 25 Feb 2024 00:13:17 GMT
User-Agent: GStreamer/1.20.6


2024/02/25 00:13:17 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 1
Public: DESCRIBE, ANNOUNCE, SETUP, PLAY, RECORD, PAUSE, GET_PARAMETER, TEARDOWN
Server: gortsplib


Progress: (open) Opened Stream
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Progress: (request) Sending RECORD request
Redistribute latency...
  5680.627656 s:  VX_ZONE_WARNING:[ownLogRtTraceAddEventClass:220] Log RT event 281472890162896 of event class 3 already exists, not adding again
  5680.628045 s:  VX_ZONE_WARNING:[ownLogRtTraceAddEventClass:220] Log RT event 281472890162897 of event class 3 already exists, not adding again
2024/02/25 00:13:17 DEB [RTSP] [conn [::1]:34110] [c->s] ANNOUNCE rtsp://localhost:554/test RTSP/1.0
CSeq: 2
Content-Length: 323
Content-Type: application/sdp
Date: Sun, 25 Feb 2024 00:13:17 GMT
User-Agent: GStreamer/1.20.6

v=0
o=- 961874131 1 IN IP6 ::1
s=Session streamed with GStreamer
i=rtspclientsink
t=0 0
a=tool:GStreamer
m=video 0 RTP/AVP 96
c=IN IP6 ::
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1;sprop-parameter-sets=Z0IACukAUYHry8IAAAMAAgAAAwB5CA==,aMqPLA==
a=control:stream=0
a=ts-refclk:local
a=mediaclk:sender

2024/02/25 00:13:17 INF [RTSP] [session 4d91d3e7] created by [::1]:34110
Progress: (record) Sending server stream info
2024/02/25 00:13:17 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 2
Server: gortsplib


Progress: (request) SETUP stream 0
Redistribute latency...
Redistribute latency...
2024/02/25 00:13:17 DEB [RTSP] [conn [::1]:34110] [c->s] SETUP rtsp://localhost:554/test/stream=0 RTSP/1.0
CSeq: 3
Date: Sun, 25 Feb 2024 00:13:17 GMT
Transport: RTP/AVP/UDP;unicast;client_port=43540-43541;mode=RECORD
User-Agent: GStreamer/1.20.6


  5680.718497 s:  VX_ZONE_WARNING:[ownLogRtTraceAddEventClass:220] Log RT event 281472890162896 of event class 3 already exists, not adding again
  5680.723527 s:  VX_ZONE_WARNING:[ownLogRtTraceAddEventClass:220] Log RT event 281472890162897 of event class 3 already exists, not adding again
2024/02/25 00:13:18 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 3
Server: gortsplib
Session: d229ae9e6ec844f1afd53048aa85f722
Transport: RTP/AVP;unicast;client_port=43540-43541;server_port=8000-8001


Progress: (record) Starting recording
2024/02/25 00:13:18 DEB [RTSP] [conn [::1]:34110] [c->s] RECORD rtsp://localhost:554/test RTSP/1.0
CSeq: 4
Date: Sun, 25 Feb 2024 00:13:18 GMT
Session: d229ae9e6ec844f1afd53048aa85f722
User-Agent: GStreamer/1.20.6


2024/02/25 00:13:18 INF [RTSP] [session 4d91d3e7] is publishing to path 'test', 1 track (H264)
2024/02/25 00:13:18 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 4
Server: gortsplib
Session: d229ae9e6ec844f1afd53048aa85f722


2024/02/25 00:13:18 DEB [RTSP] [conn 192.168.10.133:58145] [s->c] RTSP/1.0 200 OK
CSeq: 2
Content-Base: rtsp://192.168.10.122:554/test/
Content-Length: 310
Content-Type: application/sdp
Server: gortsplib

v=0
o=- 0 0 IN IP4 127.0.0.1
s=Session streamed with GStreamer
c=IN IP4 0.0.0.0
t=0 0
m=video 0 RTP/AVP 96
a=control:rtsp://192.168.10.122:554/test/trackID=0
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; profile-level-id=42000A; sprop-parameter-sets=Z0IACukAUYHry8IAAAMAAgAAAwB5CA==,aMqPLA==

Redistribute latency...
2024/02/25 00:13:18 DEB [RTSP] [conn 192.168.10.133:58145] [c->s] SETUP rtsp://192.168.10.122:554/test/trackID=0 RTSP/1.0
CSeq: 3
Transport: RTP/AVP/UDP;unicast;client_port=20860-20861
User-Agent: Mozilla/5.0


2024/02/25 00:13:18 INF [RTSP] [session af1c4f83] created by 192.168.10.133:58145
2024/02/25 00:13:18 DEB [RTSP] [conn 192.168.10.133:58145] [s->c] RTSP/1.0 200 OK
CSeq: 3
Server: gortsplib
Session: 1a373c0c69e74c6dad6e3f4790a2a277;timeout=60
Transport: RTP/AVP;unicast;client_port=20860-20861;server_port=8000-8001


2024/02/25 00:13:18 DEB [RTSP] [conn 192.168.10.133:58145] [c->s] PLAY rtsp://192.168.10.122:554/test/ RTSP/1.0
CSeq: 4
Range: npt=0.000-
Session: 1a373c0c69e74c6dad6e3f4790a2a277
User-Agent: Mozilla/5.0


2024/02/25 00:13:18 INF [RTSP] [session af1c4f83] is reading from path 'test', with UDP, 1 track (H264)
2024/02/25 00:13:18 DEB [RTSP] [conn 192.168.10.133:58145] [s->c] RTSP/1.0 200 OK
CSeq: 4
RTP-Info: url=rtsp://192.168.10.122:554/test/trackID=0;seq=4065;rtptime=1335497194
Server: gortsplib
Session: 1a373c0c69e74c6dad6e3f4790a2a277;timeout=60


Redistribute latency...
2024/02/25 00:13:23 INF [RTSP] [conn 192.168.10.133:58145] closed: EOF
2024/02/25 00:14:13 DEB [RTSP] [conn [::1]:34110] [c->s] GET_PARAMETER rtsp://localhost:554/test RTSP/1.0
CSeq: 5
Date: Sun, 25 Feb 2024 00:14:13 GMT
Session: d229ae9e6ec844f1afd53048aa85f722
User-Agent: GStreamer/1.20.6


2024/02/25 00:14:13 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 5
Content-Type: text/parameters
Server: gortsplib
Session: d229ae9e6ec844f1afd53048aa85f722


2024/02/25 00:14:22 INF [RTSP] [session af1c4f83] destroyed: session timed out
2024/02/25 00:14:27 INF [path test] runOnDemand command stopped: not needed by anyone
handling interrupt.
Interrupt: Stopping pipeline ...
Execution ended after 0:01:09.624491730
Setting pipeline to NULL ...
2024/02/25 00:14:27 DEB [RTSP] [conn [::1]:34110] [c->s] PAUSE rtsp://localhost:554/test RTSP/1.0
CSeq: 6
Date: Sun, 25 Feb 2024 00:14:27 GMT
Session: d229ae9e6ec844f1afd53048aa85f722
User-Agent: GStreamer/1.20.6


2024/02/25 00:14:27 DEB [RTSP] [conn [::1]:34110] [s->c] RTSP/1.0 200 OK
CSeq: 6
Server: gortsplib
Session: d229ae9e6ec844f1afd53048aa85f722


2024/02/25 00:14:27 INF [RTSP] [conn [::1]:34110] closed: EOF
2024/02/25 00:14:27 DEB [path test] destroyed: not in use
2024/02/25 00:14:27 INF [RTSP] [session 4d91d3e7] destroyed: not in use
Freeing pipeline ...
  5749.914627 s:  VX_ZONE_INIT:[tivxHostDeInitLocal:110] De-Initialization Done for HOST !!!
  5749.915700 s:  VX_ZONE_INIT:[tivxDeInitLocal:193] De-Initialization Done !!!
APP: Deinit ... !!!
REMOTE_SERVICE: Deinit ... !!!
REMOTE_SERVICE: Deinit ... Done !!!
IPC: Deinit ... !!!
IPC: DeInit ... Done !!!
MEM: Deinit ... !!!
DDR_SHARED_MEM: Alloc's: 67 alloc's of 202015118 bytes
DDR_SHARED_MEM: Free's : 67 free's  of 202015118 bytes
DDR_SHARED_MEM: Open's : 0 allocs  of 0 bytes
MEM: Deinit ... Done !!!
APP: Deinit ... Done !!!

^C2024/02/25 00:21:32 INF shutting down gracefully
2024/02/25 00:21:32 INF [SRT] listener is closing
2024/02/25 00:21:32 INF [WebRTC] listener is closing
2024/02/25 00:21:32 INF [HLS] listener is closing
2024/02/25 00:21:32 INF [RTMP] listener is closing
2024/02/25 00:21:32 INF [RTSP] listener is closing
2024/02/25 00:21:32 DEB path manager is shutting down
2024/02/25 00:21:32 INF waiting for running hooks
```

I'm not sure why it is keeping the session open after receiving "2024/02/25 00:13:23 INF [RTSP] [conn 192.168.10.133:58145] closed: EOF"

I can't start streaming from the image sensor successfully again unless I let it timeout, which is unwanted behavior for me. Perhaps I can easily change the Session Timeout time (perhaps just as a test)?  
Or is this perhaps due to AgentDVR not sending the proper message to disconnect and close the session? 
I was (perhaps incorrectly) expecting an RTSP TEARDOWN but didn't see one in Wireshark using an rtsp filter. AgentDVR is able to stop and start as fast as I can manually enable and disable the stream using another Onvif image sensor product however.
I also don't experience this issue using Gstreamer's  gst-rtsp-server library's test example instead but getting away from that is why I'm evaluating MediaMTX.

Thank you for your support

**Comment by cubisttriangle** [comment]
Yes it is because the the RTSP session is not being torn down correctly, so it's keeping the listener around in case of a potential reconnect. This timeout duration is hardcoded in  some library  (gortsplib? cant remember, its been a while) and not configurable via mediamtx.

One way around this is to create some IPC "callback" in the runOnUndemand hook. There, you can notify your application that all stream consumers have abandoned your stream and decide if you want to tear down the RTSP session ahead of the timeout.

**Reply by TrueDenRyu** [reply]
Thank you for your assistance. I've been trying to wrap my brain around what that might entail and haven't come up with anything to try yet. Would you or someone be able to provide an example/hint of what to put in the runOnUnDemand hook or further explain what an IPC "callback" would look like?

I'm currently thinking that this would be something like knowing the process ID generated from the runOnDemand call and killing it when somehow detecting the  "2024/02/25 00:13:23 INF [RTSP] [conn 192.168.10.133:58145] closed: EOF" in MediaMTX. That seems likely to cause problems with the runOnDemand command closing though. As I currently understand, I'm configured to have MediaMTX be my "application" and am wanting to keep it that way, if possible.

Thank You

**Reply by cubisttriangle** [reply]
When it comes time for a hook to be run, MediaMTX will attempt execute whatever value you have configured for that hook. So whatever it is that you need to do, you could do it there with some shell command.

**Comment by TrueDenRyu** [comment]
Doing some experiments using a shell script to start and stop the gstreamer video stream I observed that the runOnUnDemand still doesn't execute until after the session timeout so this doesn't seem like a viable workaround to me currently.

I did find if I killed the gstreamer stream process manually/externally to MediaMTX that the  "readTimeout" parameter in the mediamtx.yml file is respected and destroys the session according the provided time. But then I receive system messages:
"virtio_rpmsg_bus virtio1: msg received with no recipient" which I believe is because it is trying to kill a process that doesn't exist (as it has already been killed) as I figured.

Is it not a bug that MediaMTX isn't destroying the session when receiving the  "INF [RTSP] [conn [::1]:34110] closed: EOF"?
It looked to me that this likely relates to a TCP packet with the Fin Flag set using Wireshark. Perhaps this isn't definite enough to destroy a session?

Best Regards

**Reply by cubisttriangle** [reply]
There are other hooks. Like runOnUnRead. I wouldn't call it a bug. It allows time for a reconnect in the event of a network drop.

**Reply by TrueDenRyu** [reply]
Moving the gstreamer command to  runOnRead or runOnReady both result in:
2024/03/11 14:10:37 INF MediaMTX v1.6.0
2024/03/11 14:10:37 INF configuration loaded from /opt/onvif/mediamtx.yml
2024/03/11 14:10:37 INF [RTSP] listener opened on :8554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
2024/03/11 14:10:47 INF [RTSP] [conn 192.168.10.133:54834] opened
2024/03/11 14:10:47 INF [RTSP] [conn 192.168.10.133:54834] closed: no one is publishing to path 'test'

(As a side note, I did try changing the sessionTimeout in the gortsplib>server.go file, s.sessionTimeout = 1 *2 * time.Second) and this did stop the steam like I want so I can start it again but I understand better now why this isn't a great thing to do).

Perhaps using the runOnDemand isn't the correct method either.  I can't seem to restart my stream that way which was basically the reason behind creating this discussion. It looks like the preferred way is to start up MediaMTX and then publish a stream to it. I tried this originally (using the same command I put in runOnDemand in the OP) but for reasons unknown to me I can't get VLC or AgentDVR software to display a stream and it is sending ~200-500 byte RTP packets instead of the more typical ~1500 byte packets I see when it is working, as observed with Wireshark . Any idea what might cause this?

2024/03/11 14:24:40 INF MediaMTX v1.6.0
2024/03/11 14:24:40 INF configuration loaded from /opt/onvif/mediamtx.yml
2024/03/11 14:24:40 INF [RTSP] listener opened on :8554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
2024/03/11 14:24:43 INF [RTSP] [conn [::1]:53754] opened
2024/03/11 14:24:43 INF [RTSP] [session ab013354] created by [::1]:53754
2024/03/11 14:24:44 INF [RTSP] [session ab013354] is publishing to path 'test', 1 track (H264)
2024/03/11 14:24:46 INF [RTSP] [conn 192.168.10.133:54941] opened
2024/03/11 14:24:46 INF [RTSP] [session b08ff7bc] created by 192.168.10.133:54941
2024/03/11 14:24:46 INF [RTSP] [session b08ff7bc] is reading from path 'test', with UDP, 1 track (H264)
2024/03/11 14:24:59 INF [RTSP] [session b08ff7bc] destroyed: torn down by 192.168.10.133:54941
2024/03/11 14:24:59 INF [RTSP] [conn 192.168.10.133:54941] closed: EOF

(Above I see that VLC sends a TEARDOWN which closes the session, as expected)

Best Regards

[a comment by an automated account is not shown]
